### ISSUE TITLE
hcxpcaptool should be replaced by hcxpcapngtool

### ISSUE THREAD

**Issue by orsinium** [v1s1t0r1sh3r3/airgeddon#401]
#### What is your airgeddon version?

The latest commit into master (https://github.com/v1s1t0r1sh3r3/airgeddon/commit/0cb5b627e29a61decca08172bb239d275edca5d7)

#### What is your Linux O.S. and which version?

Not relevant

#### Which is the chipset of your wireless card?

Not relevant

#### Describe the issue and the steps to reproduce it

hcxpcaptool was removed from hcxtools and replaced by hcxpcapngtool. 

https://github.com/ZerBea/hcxtools/issues/158

Steps to reproduce

1. install the latest hcxtools from master, following the official instructions.
1. Run the latest airgeddon
1. Airgeddon complaints that hcxpcaptool is missing.

**Comment by OscarAkaElvis** [comment]
Thanks, for sure this is something that we should face in the future, but for now... there is no package called hcxpcapngtool. This screenshot is from a Kali Linux. I've also checked it in Parrot Linux. Same packages.

![image](https://user-images.githubusercontent.com/5803348/95822500-56bed080-0d2c-11eb-81d5-02221f501577.png)

I think hcxpcapngtool binary right now is included on hcxtools package. So for now We'll let this as is. If in the future, there is a package called hcxpcapngtool which includes both binaries (hcxpcapngtool and hcxpcaptool), we'll modify the dependencies. Let's wait this change arrive to the packages on the repositories on the main pentest Linux distros.

Thanks for the advice.

**Comment by orsinium** [comment]
I'm using debian and mint, and in both repositories there is no `hcxtools`. So, I installed `hcxtools` and `hcxdumptool` from the source. And when installing from the source, there is already no `hcxpcaptool`.

**Comment by OscarAkaElvis** [comment]
Ok, so the change is done in some Debian based Linux and is not done in others... until it is done everywhere always are going to be affected users. But from a pragmatical point of view (to do the less changes as possible), the most pf airgeddon users are using pentest distros like Kali or Parrot security Linux. So I'll wait until changes arrive to that Linux distros to perform any change.

Thanks again for your interest.

**Comment by orsinium** [comment]
> Ok, so the change is done in some Debian based Linux and is not done in others... until it is done everywhere always are going to be affected users

There are no changes in Debian. hcxtools never was in most of the repos. It is only in arch, kali, and mac. See [install via packet manager](https://github.com/ZerBea/hcxtools#or-install-via-packet-manager-of-your-distribution) section.

**Comment by OscarAkaElvis** [comment]
Ok, i have some questions to perform changes:

1. The binary called hcxpcapngtool is able to perform the same as hcxpcaptool?

2. If no, after the new changes, is the hcxpcaptool available in any package?

3. If yes, which could be the command to replace this one?

```
if hcxpcaptool -z "${tmpdir}${standardpmkid_filename}" "${tmpdir}pmkid.pcapng" | grep -q "PMKID(s) written" 2> /dev/null; then
```

That is the unique line in airgeddon using hcxpcaptool. Do you think this can be done using hcxpcapngtool binary?

Depending of this answers, the modifications needed are quite different.

**Comment by OscarAkaElvis** [comment]
I asked directly to the authors... will see the answers:

https://github.com/ZerBea/hcxtools/issues/166

**Comment by R4z0rX** [comment]
@OscarAkaElvis hcxpcapngtool now works with the new hashcat format 22000 WPA-PBKDF2-PMKID+EAPOL.
In hcxpcaptool, the output was:
```
-z <file> : output PMKID file (hashcat hashmode -m 16800)
-Z <file> : output PMKID file (hashcat hashmode -m 16801)
```
And now, in hcxpcapngtool is:
`-o <file> : output WPA-PBKDF2-PMKID+EAPOL (hashcat -m 22000) hash file`
So, I think you can replace **-z** with **-o**. [Here](https://github.com/ZerBea/hcxtools/blob/174e454b826087a9c7273caee82ff1d78d6692d5/hcxpcapngtool.c#L4854) is the help of new hcxpcapngtool. 

Right now, I'm using this in my script:
```
# 1) The capture
hcxdumptool -i $1 --enable_status=21 --filterlist_ap=$trgt --filtermode=2 -o $capfile
# 2) The conversion
hcxpcapngtool --log=myhashes.log -o $hashes $capfile
```
Hope this help.

**Comment by OscarAkaElvis** [comment]
I'll use instead of -o the --pmkid which create the old format which is what I need. I did some tests and it works.

**Comment by R4z0rX** [comment]
Ok. But, I'm curious... Why would you want to keep the old deprecated format? Or is this just a temporary fix...?

**Comment by OscarAkaElvis** [comment]
Because in that way there is no more changes to do on code. If I change the mode, then there are a lot of things more to change for the decryption menu.

Is there any advantage using the new hash format? More performance being cracked? Or is the same?

**Comment by R4z0rX** [comment]
Oh, ok, I understand now.
And yes, it has many advantages, and the performance is just slightly better (according to some reports, and my own tests).
Advantages (very resumed):
- easy to handle (you can use hcxpcapngtool to filter "bad" hashes, etc.)
- hash file is HEX ASCII (no longer binary)
- full reuse of PBKDF2 over PMKID and EAPOL (speed improvement)
- and more...

Here are some post about this:
[https://hashcat.net/forum/thread-8986-post-47705.html#pid47705](https://hashcat.net/forum/thread-8986-post-47705.html#pid47705)
In this one there's a nice explanation (from ZerBea) about Nonce Error Correction, and the impact on hashcat speed.
[https://hashcat.net/forum/thread-8986-post-47705.html#pid47705](https://hashcat.net/forum/thread-8870-post-48008.html#pid48008)

So, maybe. in this case there's no need to update all the script. But, it'll be nice to have and update in the future. :)

**Comment by OscarAkaElvis** [comment]
Ok, I did some hashcat benchmarks creating the hash with the --pmkid and using hashcat -m 16800 and using instead the -o option and the hashcat -m 22000 and it seems is a little bit faster. Is not a big difference but is a little bit and I think it worths to do the changes. I'll change that even if is more development effort.

**Comment by OscarAkaElvis** [comment]
The first commit is done... now it is using the new hcxpcapngtool. Now developing to create the hash with new format to be cracked using hashcat -m 22000 instead the old one... just only one data missing. I need to know from which hashcat version is existing this new -m 22000 format in order to do the proper validations.

**Comment by OscarAkaElvis** [comment]
Based on [this hashcat release notes](https://hashcat.net/forum/thread-9303.html) I think it is available from hashcat 6.0.0 so we can proceed now with the needed changes to support this new kind of hash (WPA-PBKDF2-PMKID+EAPOL) doing the proper validations to keep robustness.

**Comment by OscarAkaElvis** [comment]
Done... it was easy! this is the commit: https://github.com/v1s1t0r1sh3r3/airgeddon/commit/379afedb619701ef96b51b921e8d249b5dcd4d86

Now we can use all the new stuff, hcxpcapngtool and the new hash format. Yeah! 😎 

All of this stuff will be available on next v10.31. It can be tested git cloning dev branch. I tested it and it works like a charm. Closing.

**Comment by R4z0rX** [comment]
That's great! 😃
I'll give it a try later.
Thank you.

**Comment by ZerBea** [comment]
@OscarAkaElvis , the big difference between 16800/2500 and 22000 is:
If you have 2 networks using the same ESSID (let's name it "default" for this example) and you captured a PMKID from network one and an EAPOL handshake from network two, you have to run hashcat twice: 2500 for the network with the EAPOL handshake and 16800 for the network with the PMKID. That means twice times calculating a PMK by a slow PBKDF2 algo.

Running 22000 you will get both of them for the price of one. The PMK is calculated only once (by slow PBKDF2). Than it is compared to the PMKID of network one and the EAPOL handshkake of network two.

But make up your own mind:
```
network one:
$ time hashcat -m 16800 test.16800 digits
hashcat (v6.1.1-116-gfb219e0a6) starting...
real	0m4,875s
user	0m0,395s
sys	0m0,903s

network two:
$ time hashcat -m 2500 test.hccapx digits
hashcat (v6.1.1-116-gfb219e0a6) starting...
real	0m5,896s
user	0m0,421s
sys	0m0,938s

both networks inside a single hash file:
$ time hashcat -m 22000 test.22000 digits
hashcat (v6.1.1-116-gfb219e0a6) starting...
real	0m7,917s
user	0m0,469s
sys	0m1,048s
```
The same applies to JtR (hcxpcapngtool --john)

In addition to that, hcxpcapngtool is able to detect zeroed PMKIDs and PMKIDs/EAPOL messages calculated using a zeroed PMK. This ones are useless to retrieve the PSK.
Once converted from pcapng/pcap/cap (that include gz compressed dump files) to hash format 22000, you can use hcxhashtool to filter what ever you want.

BTW:
If you really need to filter the dump file, please consider to use tshark instead of wpaclean. wpaclean remove important frames from the dump file (the same applies to besside-ng).
`$ tshark -r dump.cap -R "(wlan.fc.type_subtype == 0x00 || wlan.fc.type_subtype == 0x02 || wlan.fc.type_subtype == 0x04 || wlan.fc.type_subtype == 0x05 || wlan.fc.type_subtype == 0x08 || eapol)" -2 -w filtered.pcapng
`
This will not remove frames which  possible include a PSK in the clear.
Please notice:
Default format of Wireshark/tshark is pcapng.
If you need old (cap/pcap) format, you must specify it by -F option!

**Comment by ZerBea** [comment]
Here is an example that let you imagine what you are missing if this frames are filtered out or not captured:
https://github.com/evilsocket/pwnagotchi/issues/835#issuecomment-598597214

**Comment by OscarAkaElvis** [comment]
Don't worry. I did all the needed changes. Even I started to use the new -m 22000 hash format on this commit: https://github.com/v1s1t0r1sh3r3/airgeddon/commit/379afedb619701ef96b51b921e8d249b5dcd4d86

Yeah, the "trick" of getting cap from the pcapng file using tshark was already implemented one version ago, the user is prompted to get it if he/she wants.

Thanks.

**Comment by ZerBea** [comment]
Thanks for your answer. I'm glad to hear this.
